### PR TITLE
[SYCL][CUDA] Map maxrregcount build option to a Cuda JIT option for use with ptxjit

### PR DESCRIPTION
**Pull request by GeorgeWeb**
This PR maps the "--maxrregcount" (from ptxas) compiler option string token to a Cuda JIT option (`CUjit_option`) for use with `ptxjitcompiler` to limit register pressure.

The idea is that for now, while not a complete solution, this can help edge-cases where the use of too many registers not spilling above certain threshold needs to be handled for the kernel to even be able to launch.

It is related to a PR that handles this edge case by propagating the back-end API error and throwing a targeted SYCL exception: #9106. It is also one possible solution to the jit binary from issue #8414 and to one of the issues found in the Blender project.

This string can currently be passed via the `SYCL_PROGRAM_COMPILE_OPTIONS` environment variable when executing the binary, i.e.:
```
SYCL_PROGRAM_COMPILE_OPTIONS="--maxrregcount=64" ./a.out
```

Ideally, compiler front-end to runtime propagation would make this better rather than using the `env`, but the mapping to the appropriate `CUjit_option` still needs to be handled similarly in the Cuda PI.

---

Still figuring out some appropriate tests for this change. It does fix the kernel launching errors mentioned in the above issues.

### CODE REVIEW

**Comment by jinz2014** [comment]
All is needed for the original issue is that the compiler reports to a programmer that the kernels fails to launch in the SYCL program because of using too many resources (e.g. registers).  Is that right ? 

If not, then the compiler may produce a warning message about the GPU register usage before and after some compiler optimization.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 1]
```diff
@@ -390,6 +390,37 @@ void getUSMHostOrDevicePtr(PtrT usm_ptr, CUmemorytype *out_mem_type,
...
+  }
+
+  bool valid = false;
+  const std::size_t delimPos = build_options.find('=', optionPos + 1u);
+  const std::size_t startPos = delimPos + 1;
+  if (delimPos != std::string::npos && startPos < build_options.length()) {
+    // validation function helper for value characters
+    static auto is_digit = [](char c) -> bool {
+      return std::isdigit(static_cast<unsigned char>(c)) != 0;
+    };
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:408]: This really doesn't seem to shorten the use much. Could we just inline it (and remove is `!= 0` check)?

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 2]
```diff
@@ -390,6 +390,37 @@ void getUSMHostOrDevicePtr(PtrT usm_ptr, CUmemorytype *out_mem_type,
...
+
+    std::size_t pos = startPos;
+    while (build_options[pos] != '\n' && is_digit(build_options[pos])) {
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:411]: Why is the newline check needed? Would `is_digit` consider it a digit? Instead we should check that `pos` doesn't exceed the string so we don't start reading out of bounds.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 3]
```diff
@@ -390,6 +390,37 @@ void getUSMHostOrDevicePtr(PtrT usm_ptr, CUmemorytype *out_mem_type,
...
+  using namespace std::string_view_literals;
+  const std::size_t optionPos = build_options.find_first_of("maxrregcount"sv);
+  if (optionPos == std::string::npos) {
+    return false;
+  }
+
+  bool valid = false;
+  const std::size_t delimPos = build_options.find('=', optionPos + 1u);
+  const std::size_t startPos = delimPos + 1;
+  if (delimPos != std::string::npos && startPos < build_options.length()) {
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:404]: Nit; personally I'd prefer just doing an early exit here instead. Not blocking though.

File: sycl/plugins/cuda/pi_cuda.cpp [hunk 4]
```diff
@@ -390,6 +390,37 @@ void getUSMHostOrDevicePtr(PtrT usm_ptr, CUmemorytype *out_mem_type,
...
+      return std::isdigit(static_cast<unsigned char>(c)) != 0;
+    };
+
+    std::size_t pos = startPos;
+    while (build_options[pos] != '\n' && is_digit(build_options[pos])) {
+      pos++;
+    }
+    const std::string valueString =
+        build_options.substr(startPos, pos - startPos);
+    if (!valueString.empty()) {
```
Comment (steffenlarsen) [inline sycl/plugins/cuda/pi_cuda.cpp:416]: Nit; Likewise, if you just exit here you can avoid the need for `valid` entirely.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Comment by steffenlarsen** [comment]
Failures in CI should have been addressed a while ago. @GeorgeWeb - Could you please push a merge commit for good measure?

**Pull request merged**